src/dyncfg-launchdarkly/src/lib.rs

```diff
@@ -49,41 +49,67 @@ where
-                tracing::info!("SyncedConfigSet initialize on boot: initialize has timed out");
-                None
+
+    // Attempts to return an initiated ld_client.
+    // If no key is provided None will be returned.
+    // If the client fails to build an error will be returned.
+    // If the client hangs, or cannot connect in time, a timeout error will be logged
+    // and None will be returned.
+    async fn get_ld_client(
+        key: Option<&str>,
+        config_sync_timeout: Duration,
+    ) -> Result<Option<Client>, anyhow::Error> {
+        if let Some(key) = key {
+            let init = async {
+                let init_client = ld::Client::build(ld::ConfigBuilder::new(key).build())?;
+                // You must call a start function, such as
+                // `start_with_with_default_exector` before
+                // initialized_async, or there will be no permits
+                // on the internal semaphore and the client will hang
+                // indefinitely!
+                init_client.start_with_default_executor();
+                let max_backoff = Duration::from_secs(60);
+                let mut backoff = Duration::from_secs(5);
+                while !init_client.initialized_async().await {
+                    tracing::error!("SyncedConfigSet failed to initialize");
+                    tokio::time::sleep(backoff).await;
+                    backoff = (backoff * 2).min(max_backoff);
+                }
+                Ok(init_client)
+            };
+            match mz_ore::future::timeout(config_sync_timeout, init).await {
+                Ok(client) => Ok(Some(client)),
+                Err(TimeoutError::Inner(e)) => Err(e),
+                Err(TimeoutError::DeadlineElapsed) => {
+                    tracing::info!("SyncedConfigSet initialize on boot: initialize has timed out");
+                    Ok(None)
+                }
             }
+        } else {
+            Ok(None)
         }
-    } else {
-        None
-    };
+    }
+
+    let ld_client = get_ld_client(launchdarkly_sdk_key, config_sync_timeout).await?;
 
-    let synced = SyncedConfigSet {
+    let mut synced = SyncedConfigSet {
         set,
         ld_client,
         ld_ctx: ld_ctx(build_info, ctx_builder)?,
         on_update,
     };
     synced.sync()?;
-    task::spawn(
-        || "SyncedConfigSet sync_loop",
-        synced.sync_loop(config_sync_loop_interval),
-    );
+    let launchdarkly_sdk_key = launchdarkly_sdk_key.clone().map(|k| k.to_string());
+    task::spawn(|| "SyncedConfigSet sync_loop", async move {
+        if synced.ld_client.is_none() {
+            synced.update_client(
+                get_ld_client(launchdarkly_sdk_key.as_deref(), config_sync_timeout)
+                    .await
+                    .expect("Already validate that building a client won't fail"),
+            );
+        }
+        synced.sync_loop(config_sync_loop_interval).await;
+    });
```

Comment:
Having trouble following this logic. `ld_client` might be `None` because there's no LD key, but now it looks like we're going to try to initialize one anyway—and also the `expect` here doesn't seem safe because the client might fail to initialize in 30s? 

I think this all gets simpler if you don't try to recreate the client after a timeout. Probably easiest to explain in code, 1s!

Reply:
Ah, ok I was assuming we'd need `initialized_async` to work before passing to the sync loop. 

If that's not the case this is much simpler.

Reply:
> I think the expect is mostly ok, the only actual error being returned was a build error which should be deterministic and we were already checking for it.

Ah, you're right! I got thrown off because the `init` future returns a `Result`, and so it looked like the `mz_ore::future::timeout` function could return `TimeoutError::Inner`, which would trigger the panic. Left a comment about cleaning that up below.